framework/src/Volo.Abp.AspNetCore.Mvc/Volo/Abp/AspNetCore/Mvc/Localization/AbpLanguagesController.cs

```diff
@@ -1,18 +1,42 @@
-﻿using Microsoft.AspNetCore.Localization;
+﻿using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Localization;
 using Microsoft.AspNetCore.Mvc;
+using Volo.Abp.Settings;
 
 namespace Volo.Abp.AspNetCore.Mvc.Localization
 {
     [Area("Abp")]
     [Route("Abp/Languages/[action]")]
     public class AbpLanguagesController : AbpController
     {
+        public SettingManager SettingManager { get; set; }
+
         [HttpGet]
-        public IActionResult Switch(string culture, string uiCulture = "") //TODO: Implement return URL
+        public IActionResult Switch(string culture, string uiCulture = "", string returnUrl = "")
         {
-            //TODO: Check allowed languages and so on...
+            if (!GlobalizationHelper.IsValidCultureCode(culture))
+            {
+                throw new AbpException("Unknown language: " + culture + ". It must be a valid culture!");
+            }
+
+            if (!GlobalizationHelper.IsValidCultureCode(uiCulture))
+            {
+                throw new AbpException("Unknown language: " + uiCulture + ". It must be a valid culture!");
+            }
+
+            string cookieValue = CookieRequestCultureProvider.MakeCookieValue(new RequestCulture(culture, uiCulture));
+
+            Response.Cookies.Append(CookieRequestCultureProvider.DefaultCookieName, cookieValue, new CookieOptions
+            {
+                Expires = Clock.Now.AddYears(2),
+                HttpOnly = true
+            });
+
+            if (!string.IsNullOrWhiteSpace(returnUrl))
+            {
+                return Redirect(returnUrl);
```

Comment:
Return URL should be checked for security. Example: https://github.com/abpframework/abp/blob/master/modules/account/src/Volo.Abp.Account.Web/Pages/Account/Login.cshtml.cs#L70